src/Microsoft.ML.Data/DataLoadSave/PartitionedFileLoader.cs

```diff
@@ -364,29 +364,35 @@ private sealed class Cursor : RootCursorBase
         {
             private PartitionedFileLoader _parent;
 
-            private bool[] _active;
-            private bool[] _subActive; // Active columns of the sub-cursor.
+            private readonly bool[] _active;
+            private readonly bool[] _subActive; // Active columns of the sub-cursor.
             private Delegate[] _getters;
             private Delegate[] _subGetters; // Cached getters of the sub-cursor.
 
+            private readonly IEnumerable<Schema.Column> _colsNeeded;
+            private readonly IEnumerable<Schema.Column> _subActiveColsNeeded;
+
             private ReadOnlyMemory<char>[] _colValues; // Column values cached from the file path.
             private RowCursor _subCursor; // Sub cursor of the current file.
 
             private IEnumerator<int> _fileOrder;
 
-            public Cursor(IChannelProvider provider, PartitionedFileLoader parent, IMultiStreamSource files, Func<int, bool> predicate, Random rand)
+            public Cursor(IChannelProvider provider, PartitionedFileLoader parent, IMultiStreamSource files, IEnumerable<Schema.Column> colsNeeded, Random rand)
                 : base(provider)
             {
                 Contracts.AssertValue(parent);
                 Contracts.AssertValue(files);
-                Contracts.AssertValue(predicate);
 
                 _parent = parent;
+                _colsNeeded = colsNeeded;
 
-                _active = Utils.BuildArray(Schema.Count, predicate);
+                _active = Utils.BuildArray(Schema.Count, colsNeeded);
                 _subActive = _active.Take(SubColumnCount).ToArray();
                 _colValues = new ReadOnlyMemory<char>[Schema.Count - SubColumnCount];
 
+                // sefilipi: Test this
+                _subActiveColsNeeded = Schema.Where(x => (_subActive?.Length > x.Index) && _subActive[x.Index]);
```

Comment:
>_subActiveColsNeeded [](start = 16, length = 20)

is this same as `Utils.BuildArray(SubColumCount, colsNeeded);` ?

Reply:
that returns an array, this is an IEnumerable<Schema.Column>

---
In reply to: [246980849](https://github.com/dotnet/machinelearning/pull/1975#discussion_r246980849) [](ancestors = 246980849)